lucene/core/src/java/org/apache/lucene/search/ParallelSortedCollector.java

```diff
@@ -0,0 +1,612 @@
+   * possible while avoiding thread contention on this synchronized method. A good heuristic seems to be the least power
+   * of 2 greater than than the number of threads in use.</p>
+   */
+  static class MaxScoreTerminator {
+
+    // This strategy performs best for higher N. The threshold below was determined empirically. The best tradeoff is
+    // probably related to number of threads used by the searcher and the overall counting threshold.
+    private static final int TERMINATION_STRATEGY_HIT_THRESHOLD = 50;
+
+    // By default we use 2^5-1 to check the remainder with a bitwise operation, but for best performance
+    // the actual value should always be set by calling setIntervalBits()
+    private static final int DEFAULT_INTERVAL_BITS = 5;
+    private int intervalMask;
+    int interval;
+
+    /** The worst score for each leaf */
+    private final List<LeafState> leafStates = new ArrayList<>();
+    private final List<LeafState> scratch = new ArrayList<>();
+
+    /** How many hits were requested: Collector's numHits. */
+    private final int numHits;
+
+    /** the worst hit over all */
+    private final LeafState thresholdState;
+
+    /** The total number of docs to collect: the max of the Collector's numHits and its early termination threshold. */
+    final int totalToCollect;
+
+    /** An upper bound on the number of docs "excluded" from max-score accounting due to early termination. */
+    private int numExcludedBound;
+
+    /** A lower bound on the total hits collected by all leaves */
+    private volatile int totalCollected;
+
+    /**
+     * @param numHits the number of hits to be returned
+     * @param collectionThreshold collect at least this many, even if numHits is less, for the sake of counting
+     * @param numThreads the number of threads that will be updating this concurrently. Controls the update interval.
+     *                   If null, an internal default is used.
+     */
+    MaxScoreTerminator(int numHits, int collectionThreshold, Integer numThreads) {
+      this.numHits = numHits;
+      this.totalToCollect = Math.max(numHits, collectionThreshold);
+      thresholdState = new LeafState();
+      thresholdState.score = Double.MAX_VALUE;
+      thresholdState.docid = Integer.MAX_VALUE;
+      if (numThreads != null) {
+        int numThreadsLog2 = 31 - Integer.numberOfLeadingZeros(numThreads);
+        setIntervalBits(numThreadsLog2 + 1);
+      } else {
+        setIntervalBits(DEFAULT_INTERVAL_BITS);
+      }
+    }
+
+    /**
+     * @param sort the query sort
+     * @param numHits the number of hits requested
+     * @return whether the Sort is compatible with early termination using {@link MaxScoreTerminator} and the number of hits
+     * requested is large enough to make this optimization likely to help.
+     */
+    static boolean isApplicable(Sort sort, int numHits) {
+      return numHits > TERMINATION_STRATEGY_HIT_THRESHOLD && sortIsNumeric(sort);
```

Comment:
`intervalMask` could be used as a heuristic (`numHits > interval x N`)  ?

Reply:
Oh I like it!